handbook/python-dataplugins/python-dataplugins.md

```diff
@@ -0,0 +1,308 @@
+            "unit_string": "km/h",
+            "type": "DataTypeChnFloat64"
+         }]
+   }]
+}
+
+file_path = os.path.realpath(parameter["file"])
+return {Path(file_path).stem: self.tdm_tree}
+```
+
+</details>
+
+<!-- markdownlint-disable -->
+<details>
+<summary>Dictionary Schema</summary>
+<!-- markdownlint-enable -->
+
+```python
+import datetime
+from schema import And, Schema
+
+Schema({
+      Optional('author'): str,
+      Optional('description'): str,
+      'groups': [{
+         'name': str,
+         Optional('description'): str,
+         Optional('time'): datetime.datetime,
+         'channels': [{
+            'name': str,
+            Optional('description'): str,
+            'values': list,
+            Optional('unit_string'): str,
+            'type': And(str, lambda s: s in (
+               'DataTypeChnFloat32',
+               'DataTypeChnFloat64',
+               'DataTypeChnString',
+               'DataTypeChnDate',
+               'DataTypeChnUInt8',
+               'DataTypeChnInt16',
+               'DataTypeChnInt32',
+               'DataTypeChnInt64'))
+            }]
+      }]}, ignore_extra_keys=True)
+```
+
+All further extra keys will show up as custom properties in DIAdem, Labview or
+ SystemLink DataFinder.
+
+</details>
+
+See full example: [csv-read-with-direct-loading](https://github.com/ni/systemlink-operations-handbook/tree/master/examples/python-dataplugin-examples/csv-read-with-direct-loading)
+
+## Callback Loading
+
+When handling a big data set, you might not want to load all data at the same time.
+To ensure the DataPlugin only returns values that applications request, do ...
+
+   1. assign an empty array to the channel values property:
+
+      ```python
+      ...
+      'groups': [{
+         ...
+         'channels': [{
+            ...
+            'values': [],
+         }]
+      }]
+      ```
+
+   1. outsource the functionality to load the bulk data values in a separate function.
+   The function has the following definition:
+
+      ```python
+      def read_channel_values(self, grp_index, chn_index, numberToSkip, numberToTake):
+         """
+            Returns a value array of the correct data type specified in the tdm dictionary
+         """
+      ```
+
+<!-- markdownlint-disable -->
+<details>
+<summary>Example Code</summary>
+<!-- markdownlint-enable -->
+
+```python
+def read_channel_values(self, grp_index, chn_index, numberToSkip, numberToTake):
+   dataType = self.tdm_tree["groups"][grp_index]["channels"][chn_index]["type"]
+   values = []
+   for row in self.data:
+      value = row[self.channelNames[chn_index]]
+      values.append(value)
+   return values[numberToSkip:numberToTake+numberToSkip]
+```
+
+</details>
+
+The client applications are calling that function to retrieve channel values (or
+a subset of values). The DataPlugin needs to implement the function to return the
+values for a given group and channel index. It also needs to ensure only the

```

Comment:
How does the DP ensure the correct subset of values is returned?

Reply:
changed it, see above